# Cleaning and analyzing used car listings from German eBay

For this project, we will be using a dataset that was originally scraped and uploaded to Kaggle by user orgesleka. The original dataset isn't available on Kaggle anymore, but can be found [here](https://data.world/data-society/used-cars-data). The dataset we are using contains 50,000 data points from the full dataset, and has been dirtied a bit to allow for data cleaning practice.

In [13]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv",encoding="Latin-1")

In [14]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [15]:
autos.info()
autos.head(5)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can see that the dataset has 50,000 entries (as expected) and 20 columns. However not all entries have info about the vehicle type, transmission type (gearbox), car model name, type of fuel the car uses and damage to the car that has not been repaired. The dataset also contains a combination of object and integer data types, with most columns having objects.
The column names use camelcase instead of Python's preferred snakecase.

## Converting column names

To start with, we are going to convert the column names from camelcase to snakecase and make some of them more descriptive.

In [16]:
headers = list(autos.columns.values)
print(headers)

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model', 'odometer', 'monthOfRegistration', 'fuelType', 'brand', 'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode', 'lastSeen']


First, we'll replcaed 4 of the column names with phrases that make it clearer what info the columns contain. This is to make the dataset easier to work with.

In [17]:
# Replacing some headers with more descriptive names
headers[7] = 'registration_year'
headers[12] = 'registration_month'
headers[15] = 'unrepaired_damage'
headers[16] = 'ad_created'
print(headers)

['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest', 'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model', 'odometer', 'registration_month', 'fuelType', 'brand', 'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode', 'lastSeen']


Then we'll convert the rest of the names from camelcase to snakecase to make them easier to work with in Python. I have 2 methods for doing this - using regex, and using a simple Python function.

In [18]:
# Converting camelcase to snakecast
import re

for name in headers:
    snakecase_name = re.sub(r'(?<!^)(?=[A-Z])', '_', name).lower()
    position = headers.index(name)
    headers[position] = snakecase_name
print(headers)

['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest', 'vehicle_type', 'registration_year', 'gearbox', 'power_p_s', 'model', 'odometer', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code', 'last_seen']


In [19]:
# Alternative for converting camelcase to snakecase without regex

# First, we initialize a variable ‘res’ with an empty list and append first character (in lower case) to it. 
# Now, Each time we encounter a Capital alphabet, we append ‘_’ and the alphabet (in lower case) to ‘res’, otherwise, just append the alphabet only.  
# Python3 program to convert string
# from camel case to snake case
 
# def change_case(str):
#     res = [str[0].lower()]
#     for c in str[1:]:
#         if c in ('ABCDEFGHIJKLMNOPQRSTUVWXYZ'):
#             res.append('_')
#             res.append(c.lower())
#         else:
#             res.append(c)
     
#     return ''.join(res)
     
# # Driver code
# for name in headers:
#     snakecase_name = change_case(name)
#     position = headers.index(name)
#     headers[position] = snakecase_name
# print(headers)

In [20]:
# Renaming columns in dataframe with snakecase list
autos.columns = headers
autos.head(5)

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Exploring and cleaning the data

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

In [21]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-10 15:36:24  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

From the descriptive statistics about the columns in the dataset, we can see that the following columns that have mostly one value and can be dropped:
- Seller
- Offer type
- abtest
- gearbox
- unrepaired damage
- nr_of_pictures

The following columns also need more investigation as they have 'NaN' for unique values, top value and the frequency of the top value.
- registration_year
- power_p_s
- registration_month
- postal_code

In [22]:
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

In [23]:
autos['power_p_s'].value_counts()

0        5500
75       3171
60       2195
150      2046
140      1884
         ... 
650         1
490         1
362         1
153         1
16312       1
Name: power_p_s, Length: 448, dtype: int64

In [24]:
autos['registration_month'].value_counts()

0     5075
3     5071
6     4368
5     4107
4     4102
7     3949
10    3651
12    3447
9     3389
11    3360
1     3282
8     3191
2     3008
Name: registration_month, dtype: int64

In [25]:
autos['postal_code'].value_counts()

10115    109
65428    104
66333     54
45888     50
44145     48
        ... 
23942      1
83365      1
95683      1
97794      1
67585      1
Name: postal_code, Length: 7014, dtype: int64

Further exploration of these columns reveals the following:
- There are values in registration_year that are not realistic e.g. 1001 and 9996, and should be removed as cars had not been invented in the former year and the latter has not passed yet
- There are values in power_p_s that cannot be real e.g. 16312 and should also be removed, as modern supercars have horsepower of around 500 PS
- There are 5075 entries with no month
- None of the data in the postal code column looks out of the ordinary

The following columns also contain numeric data stored as text that needs to be cleaned and converted:
- price
- odometer

Let's clean these 2 columns.

In [26]:
# Retain only numeric values
# autos['price'] = autos['price'].map(lambda x: x.lstrip('$,'))
# print(autos['price'])
autos['price'] = autos['price'].str.replace("$",'').str.replace(",",'')
autos['odometer'] = autos['odometer'].str.replace('km','').str.replace(',','')

In [27]:
# Convert the strings to integers
autos['price'] = pd.to_numeric(autos['price'])
autos['odometer'] = pd.to_numeric(autos['odometer'])

In [28]:
# Rename odometer column to indicate units
autos.columns = ['odometer_km' if x=='odometer' else x for x in autos.columns]
autos.head(2)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   

   power_p_s   model  odometer_km  registration_month fuel_type    brand  \
0        158  andere       150000                   3       lpg  peugeot   
1        286     7er       150000                   6    benzin      bmw   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08

Let's continue exploring these 2 columns for additional data that may not look right.

In [29]:
# Unique price values
autos['price'].unique().shape

(2357,)

In [30]:
# Viewing descriptive statistics about price
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [32]:
#Exploring the price column
autos['price'].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

From the exploration, we can see that the average price of a used car in the dataset is 9840. However, there are a small number of cars that have prices >100,000, 1421 cars with no price, and a few hundred with a price <100. As the car brands in the dataset do not contain the most premium brands, it's likely that we should remove entries with prices >100,000. We should also remove entries with no price and those with <100 as these are unlikely to be accurate.

In [33]:
#Filtering to a reasonable price rrange
autos_filtered_price = autos[autos["price"].between(101,99999)]

In [34]:
# Unique odometer values
autos['odometer_km'].unique().shape

(13,)

In [35]:
# Viewing descriptive statistics about price
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [36]:
#Exploring the price column
autos['odometer_km'].value_counts().sort_index(ascending=False).head(50)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

We can see that the average odometer reading for the used cars is 125,732km, which is rather high given that average drivers cover 16,000km-19,000km in a year. The 50%, 75% and max values are all also 150,000km. However, as these make up more than half the dataset, we will not remove them. They are also possible given that cars can have take up to about 320,000km with proper maintenance. The lower values for odometer readings also do not look problematic.

In [37]:
autos_filtered_price

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type  price   abtest vehicle_type  registration_year  \
0      privat    Angebot   5000  control          bus               2004   
1      privat    Angebot   8500  control    limousine               1997   
2      privat    Angebot   8990     test    limousine               2009   
3      privat    Angebot   4350  control   kleinwagen               2007   
4      privat    Angebot   1350     test        kombi               2003   
...       ...        ...    ...      ...          ...                ...   
49995  privat    Angebot  24900  control    limousine               2011   
49996  privat    Angebot   1980  control       cabrio               1996   
49997  privat    Angebot  13200     test       cabrio               2014   
49998  privat    Angebot  22900  control        kombi               2013   
49999  privat    Angebot   1250  control    limousine               1996   

         gearbox  power_p_s   model  odometer_km  registration_month  \
0        manuell        158  andere       150000                   3   
1      automatik        286     7er       150000                   6   
2        manuell        102    golf        70000                   7   
3      automatik         71  fortwo        70000                   6   
4        manuell          0   focus       150000                   7   
...          ...        ...     ...          ...                 ...   
49995  automatik        239      q5       100000                   1   
49996    manuell         75   astra       150000                   5   
49997  automatik         69     500         5000                  11   
49998    manuell        150      a3        40000                  11   
49999    manuell        101  vectra       150000                   1   

      fuel_type       brand unrepaired_damage           ad_created  \
0           lpg     peugeot              nein  2016-03-26 00:00:00   
1        benzin         bmw              nein  2016-04-04 00:00:00   
2        benzin  volkswagen              nein  2016-03-26 00:00:00   
3        benzin       smart              nein  2016-03-12 00:00:00   
4        benzin        ford              nein  2016-04-01 00:00:00   
...         ...         ...               ...                  ...   
49995    diesel        audi              nein  2016-03-27 00:00:00   
49996    benzin        opel              nein  2016-03-28 00:00:00   
49997    benzin        fiat              nein  2016-04-02 00:00:00   
49998    diesel        audi              nein  2016-03-08 00:00:00   
49999    benzin        opel              nein  2016-03-13 00:00:00   

       nr_of_pictures  postal_code            last_seen  
0                   0        79588  2016-04-06 06:45:54  
1                   0        71034  2016-04-06 14:45:08  
2                   0        35394  2016-04-06 20:15:37  
3                   0        33729  2016-03-15 03:16:28  
4                   0        39218  2016-04-01 14:38:50  
...               ...          ...                  ...  
49995           

In [38]:
autos_filtered_price.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 48051 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        48051 non-null  object
 1   name                48051 non-null  object
 2   seller              48051 non-null  object
 3   offer_type          48051 non-null  object
 4   price               48051 non-null  int64 
 5   abtest              48051 non-null  object
 6   vehicle_type        43671 non-null  object
 7   registration_year   48051 non-null  int64 
 8   gearbox             45879 non-null  object
 9   power_p_s           48051 non-null  int64 
 10  model               45681 non-null  object
 11  odometer_km         48051 non-null  int64 
 12  registration_month  48051 non-null  int64 
 13  fuel_type           44215 non-null  object
 14  brand               48051 non-null  object
 15  unrepaired_damage   39232 non-null  object
 16  ad_created          48

After filtering and cleaning, there are 2 more columns with integer values, and about 2000 entries have been removed.

## Dealing with the columns with dates 

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:

- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `ad_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website

Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas, with full timestamp values. We will use the first 10 characters (the day) to calculate the distribution of values in the date_crawled, ad_created, and last_seen columns (all string columns) as percentages.

In [44]:
# Calculating the distribution of values in date_crawled column
autos_filtered_price['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.001394
2016-04-06    0.003163
2016-04-05    0.013090
2016-04-04    0.036586
2016-04-03    0.038605
2016-04-02    0.035629
2016-04-01    0.033652
2016-03-31    0.031841
2016-03-30    0.033756
2016-03-29    0.034110
2016-03-28    0.034984
2016-03-27    0.031113
2016-03-26    0.032216
2016-03-25    0.031404
2016-03-24    0.029448
2016-03-23    0.032299
2016-03-22    0.032840
2016-03-21    0.037273
2016-03-20    0.037772
2016-03-19    0.034755
2016-03-18    0.012841
2016-03-17    0.031508
2016-03-16    0.029448
2016-03-15    0.034214
2016-03-14    0.036669
2016-03-13    0.015733
2016-03-12    0.036961
2016-03-11    0.032674
2016-03-10    0.032382
2016-03-09    0.033007
2016-03-08    0.033215
2016-03-07    0.035983
2016-03-06    0.014068
2016-03-05    0.025369
Name: date_crawled, dtype: float64

The data was scraped from the web between 5 Mar 2016 and 7 Apr 2016, with roughly equal amounts of entries scraped daily from the start date to 4 Apr 2016.

In [45]:
autos_filtered_price['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.001249
2016-04-06    0.003247
2016-04-05    0.011821
2016-04-04    0.036940
2016-04-03    0.038855
                ...   
2015-12-05    0.000021
2015-11-10    0.000021
2015-09-09    0.000021
2015-08-10    0.000021
2015-06-11    0.000021
Name: ad_created, Length: 76, dtype: float64

The ads were created between 11 Nov 2015 and 7 Apr 2016, with a larger % of ads created daily in the later part of this period.

In [46]:
autos_filtered_price['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False)

2016-04-07    0.132214
2016-04-06    0.221993
2016-04-05    0.125034
2016-04-04    0.024536
2016-04-03    0.025140
2016-04-02    0.024849
2016-04-01    0.022892
2016-03-31    0.023870
2016-03-30    0.024661
2016-03-29    0.022330
2016-03-28    0.020790
2016-03-27    0.015504
2016-03-26    0.016649
2016-03-25    0.019126
2016-03-24    0.019687
2016-03-23    0.018584
2016-03-22    0.021394
2016-03-21    0.020561
2016-03-20    0.020666
2016-03-19    0.015775
2016-03-18    0.007305
2016-03-17    0.028095
2016-03-16    0.016420
2016-03-15    0.015837
2016-03-14    0.012632
2016-03-13    0.008886
2016-03-12    0.023850
2016-03-11    0.012424
2016-03-10    0.010655
2016-03-09    0.009552
2016-03-08    0.007305
2016-03-07    0.005390
2016-03-06    0.004308
2016-03-05    0.001082
Name: last_seen, dtype: float64

The ads were last seen between 5 Mar 2016 to 7 Apr 2016, which is the same range of dates as all the scraping. However, the majority (~48%) were last seen on the last 3 days of scraping (5-7 Apr 2016), which suggests that nearly 50% of the total number of cars had not been sold by that date.

In [47]:
autos_filtered_price['registration_year'].describe()

count    48051.000000
mean      2004.749433
std         88.052996
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The mean year of registration is 2004, which makes is relatively recent for used cars. However, the minimum and maximum values for the registration year are not realistic (1000 and 9996), and should be removed as cars had not been invented in the former year and the latter has not passed yet.

Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate, so that will be our upper bound. We will use 1920 as a generous lower bound, as it is after WWI and around the start of the period of tremendous growth in automobile manufacturing in Europe [link](https://www.britannica.com/technology/automotive-industry/Growth-in-Europe).

In [48]:
# Filtering out impossible registration years and impossible power for the car
autos_filtered_year = autos_filtered_price[autos_filtered_price["registration_year"].between(1920,2016)]
autos_filtered_year = autos_filtered_year[autos_filtered_year['power_p_s'] < 1000]
autos_filtered_year['registration_year'].value_counts(normalize=True).sort_index(ascending=False).head(10)

2016    0.025773
2015    0.008150
2014    0.014306
2013    0.017254
2012    0.028287
2011    0.034985
2010    0.034270
2009    0.045064
2008    0.047817
2007    0.049161
Name: registration_year, dtype: float64

After filtering out cars registered before 1920 and after 2016, we can see that there were a very small number of cars registered sporadically before the 1980s, and most cars were registered from the 1990s onwards.

## Aggregating data by brand for comparison

In [50]:
# Getting the number of occurrences of each brand in the dataset as percentages
brand_counts = autos_filtered_year['brand'].value_counts(normalize=True).sort_values(ascending=False)
brand_counts

volkswagen        0.211601
bmw               0.110526
opel              0.107014
mercedes_benz     0.096913
audi              0.087029
ford              0.069537
renault           0.046972
peugeot           0.029870
fiat              0.025556
seat              0.018273
skoda             0.016474
nissan            0.015368
mazda             0.015217
smart             0.014241
citroen           0.014046
toyota            0.012811
hyundai           0.010058
sonstige_autos    0.009429
volvo             0.009126
mini              0.008822
mitsubishi        0.008172
honda             0.007847
kia               0.007110
alfa_romeo        0.006698
suzuki            0.005874
chevrolet         0.005679
porsche           0.005484
chrysler          0.003533
dacia             0.002666
daihatsu          0.002493
jeep              0.002298
subaru            0.002124
land_rover        0.002124
saab              0.001669
jaguar            0.001539
daewoo            0.001474
trabant           0.001366
r

We can see that there are over 40 brands in the dataset and that most of the cars are concentrated among a small number of brands. Let's focus on brands that represent at least 5% of the cars in the dataset, which make up about 65% of the cars, to see how the average price compares across these brands.

In [51]:
# Filtering the series to get only the most popular brands
top_brand = brand_counts[brand_counts >= 0.05]
top_brands = list(top_brand.index.values)
top_brands

['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford']

In [52]:
# Converting to dataframe to filter
# brand_df = brand_counts.to_frame(name='listings')
# top_brand = brand_df[brand_df['listings'] >= 0.05]
# top_brand = top_brand.reset_index()
# top_brand

In [53]:
brands = {}

for brand in top_brands:
    prices = autos_filtered_year[autos_filtered_year['brand'] == brand]
    mean_price = prices["price"].mean()
    brands[brand] = mean_price
    
print(brands)
    

{'volkswagen': 5453.471010038927, 'bmw': 8254.535595214747, 'opel': 3018.5282560259266, 'mercedes_benz': 8580.921941400135, 'audi': 9349.86301369863, 'ford': 3767.856608478803}


Of the 6 most popular brands for used cars listed on German eBay, 5 are German and one is America (Ford). Audi, Mercendes Benz and BMW are the most expensive, averaging 8000-9000 per car, while Opel and Ford are the cheapest at between 3000-4000 per car. Volkswagen is in the middle at about 5500 per car.

## Getting the average mileage per car for the top 6 brands

Let's use aggregation to understand the average mileage for those cars and if there's any visible link with mean price. While our natural instinct may be to display both aggregated series objects and visually compare them, this has a few limitations:

- It's difficult to compare more than two aggregate series objects if we want to extend to more columns
- We can't compare more than a few rows from each series object
- We can only sort by the index (brand name) of both series objects so we can easily make visual comparisons

Instead, we will create a dictionary for mean mileage too and combine all the values into a single dataframe for comparison.

In [56]:
brands_mileage = {}

for brand in top_brands:
    mileage = autos_filtered_year[autos_filtered_year['brand'] == brand]
    mean_mileage = mileage["odometer_km"].mean()
    brands_mileage[brand] = mean_mileage
    
print(brands_mileage)

{'volkswagen': 128796.86539643515, 'bmw': 132768.18984114533, 'opel': 129428.80291675107, 'mercedes_benz': 131113.8447774547, 'audi': 129313.82316313824, 'ford': 124269.01496259352}


In [81]:
mean_price_df = pd.DataFrame.from_dict(brands, orient='index', columns=['mean_price'])
# mean_price_df

In [82]:
mean_mileage_series = pd.Series(brands_mileage)
mean_mileage_list = list(mean_mileage_series.values)
mean_price_df['mean_mileage'] = mean_mileage_list
# mean_mileage_series
mean_price_df

mean_price   mean_mileage
volkswagen     5453.471010  128796.865396
bmw            8254.535595  132768.189841
opel           3018.528256  129428.802917
mercedes_benz  8580.921941  131113.844777
audi           9349.863014  129313.823163
ford           3767.856608  124269.014963

There is no visible link between mileage and price. BMW, Mercedes Benz and Audi cars have the highest average mileage (>129,000km), but are also teh most expensive. Furthermore, while Audi cars have similar average mileage to Volkswagen and Opel cars, the average price of Audis is significanly higher. Ford cars have the lowest average mileage by a bit (~120,000km) but are only the 2nd cheapest on average. 

Additional steps to consider for data cleaning:

- Identify categorical data that uses german words, translate them and map the values to their english counterparts
- Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
- See if there are particular keywords in the name column that you can extract as new columns

Additional steps to consider for analysis:
- Find the most common brand/model combinations
- Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
- How much cheaper are cars with damage than their non-damaged counterparts?